src/libsyntax/codemap.rs

```diff
@@ -592,8 +592,32 @@ impl CodeMap {
         }
     }
 
-    /// Given a `Span`, try to get a shorter span ending just after the first
-    /// occurrence of `char` `c`.
+    /// Given a `Span`, get a new `Span` covering the first token and all its trailing whitespace or
+    /// the original `Span`.
+    ///
+    /// If `sp` points to `"let mut x"`, then a span pointing at `"let "` will be returned.
+    pub fn span_until_non_whitespace(&self, sp: Span) -> Span {
+        if let Ok(snippet) = self.span_to_snippet(sp) {
+            let mut offset = 0;
+            let mut pos = 0;
+            // get the bytes width of all the non-whitespace characters
+            for (i, c) in snippet.chars().take_while(|c| !c.is_whitespace()).enumerate() {
+                offset += c.len_utf8();
+                pos = i + 1;
```

Comment:
The `pos` variable doesn't seem right. It's storing the character index but it's being used as a byte index below. The following should work I think:

```rust
let mut offset = 0;
// get the bytes width of all the non-whitespace characters
for c in snippet.chars().take_while(|c| !c.is_whitespace()) {
    offset += c.len_utf8();
}
// get the bytes width of all the whitespace characters after that
for c in snippet[offset..].chars().take_while(|c| c.is_whitespace()) {
    offset += c.len_utf8();
}
```

Reply:
You're right, I was thinking about slicing on a char slice, not the original string. Will change to your proposed code.